node/src/components/contract_runtime.rs

```diff
@@ -548,65 +567,70 @@ impl ContractRuntime {
     }
 
     pub(crate) fn set_initial_state(&mut self, sequential_block_state: ExecutionPreState) {
-        self.execution_pre_state = sequential_block_state;
+        *self.execution_pre_state.lock().unwrap() = sequential_block_state;
     }
 
-    fn execute_finalized_block<REv>(
-        &mut self,
+    #[allow(clippy::too_many_arguments)]
+    async fn execute_finalized_block_or_requeue<REv>(
+        engine_state: Arc<EngineState<LmdbGlobalState>>,
+        metrics: Arc<ContractRuntimeMetrics>,
+        exec_queue: ExecQueue,
+        execution_pre_state: Arc<Mutex<ExecutionPreState>>,
         effect_builder: EffectBuilder<REv>,
         protocol_version: ProtocolVersion,
-        execution_pre_state: ExecutionPreState,
         finalized_block: FinalizedBlock,
         deploys: Vec<Deploy>,
-    ) -> Effects<ContractRuntimeRequest>
-    where
+    ) where
         REv: From<ContractRuntimeRequest>
             + From<ContractRuntimeAnnouncement>
             + From<ControlAnnouncement>
             + Send,
     {
+        let current_execution_pre_state = execution_pre_state.lock().unwrap().clone();
         let BlockAndExecutionEffects {
             block,
             execution_results,
             maybe_step_execution_effect,
-        } = match operations::execute_finalized_block(
-            self.engine_state.as_ref(),
-            self.metrics.as_ref(),
-            protocol_version,
-            execution_pre_state,
-            finalized_block,
-            deploys,
-        ) {
+        } = match tokio::task::unconstrained(async move {
+            operations::execute_finalized_block(
+                engine_state.as_ref(),
+                metrics.as_ref(),
+                protocol_version,
+                current_execution_pre_state,
+                finalized_block,
+                deploys,
+            )
+        })
+        .await
+        {
             Ok(block_and_execution_effects) => block_and_execution_effects,
-            Err(error) => return fatal!(effect_builder, "{}", error).ignore(),
+            Err(error) => return fatal!(effect_builder, "{}", error).await,
         };
 
-        self.execution_pre_state = ExecutionPreState::from(block.header());
+        let new_execution_pre_state = ExecutionPreState::from(block.header());
+        *execution_pre_state.lock().unwrap() = new_execution_pre_state.clone();
```

Comment:
This looks like `set_initial_state`?

Reply:
That's true that it looks similar, but here we are in an `async` task that has a different lifetime, and `set_initial_state` is a method on the component. We can't use that here.